feat(cdk): adding tags parameter option to cdk deploy command

IsmaelMartinez opened the pull request:
Adding tags parameter option to cdk deploy command to allow tagging full stacks and their associated resources. 

Now it will be possible to:

```
const app = new App();

const stack1 = new Stack(app, 'stack1', { tags: { foo: 'bar' } });
const stack2 = new Stacl(app, 'stack2');

stack1.node.apply(new Tag('fii', 'bug'));
stack2.node.apply(new Tag('boo', 'bug'));
```
That will produce 
* stack1 with tags `foo bar` and `fii bug`
* stack2 with tags `boo bug`

It is possible also to override constructor tags with the stack.node.apply. 

So doing:
```
stack1.node.apply(new Tag('foo', 'newBar');
```
stack1 will have tags `foo newBar` and `fii bug`

Last, but not least, it is also possible to pass it via arguments (using yargs) as in the following example:

```
cdk deploy --tags foo=bar --tags myTag=myValue
```
That will produce a stack with tags `foo bar`and `myTag myValue`

**Important**
That will ignore tags provided by the constructor and/or aspects. 

Fixes #932 


----
### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----
Notes: 
* This is my first pull request so I will probably be missing quite a few things. Apologies for this. I will add comments in the lines were we have take some assumptions on how to do it.

* I haven't found any tests/readme around the functionality that I am extending so not added any extra tests/documentation. Do point me out to documentation/test cases that should be covered if required.

* Manually tested that the following works with a stack with a dynamoDB in it:
[x] Creating a new stack with the --tags argument creates tags for the stack and the dynamoDB resource
[x] Creating a new stack without the --tags argument does not create tags for the stack nor modifies any dynamoDB tags generated with #1516
[x] Updating a stack with a tags but not changed, correctly detects that there are no changes
[x] Updating a stack with tag values or keys changes, updated the stack tags and their associated resources.
[x] Updating a stack removing tag values, updates the stack removing the removed tags.

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

IsmaelMartinez commented:
I have unmarked the tasks because I suspect that is for the reviewer to complete/not me. Let me know if that is the actual process or not. Thanks!

eladb commented:
I think it would make more sense to apply stack-level tags in code (similar to any other tag) and propagate that information through CDK metadata to the toolkit. This feels like what people would expect and will also be more robust when it comes to multi-stack apps.

For example, I would expect this to "Just Work":

```ts
const app = new App();

const stack1 = new Stack(app, 'stack1', { tags: { foo: 'bar' } });
const stack2 = new Stacl(app, 'stack2');

stack2.node.apply(new Tag('boo', 'bug'));
```

Thew result should be:

1. All resources in `stack1`, including the stack itself have the `{ "foo": "bar" }` tag.
2. All resources in `stack2`, including the stack itself have the `{ "boo": "bug" }` tag.

@moofish32 I am curious what you think. about this.

IsmaelMartinez commented:
That sounds like the best solution. I just wanted to allow to tag the stacks without needing to use the aws cli.

What about providing both options? I will appreciate a few pointers on where the code (area, not specific) to change to allow the changes you requested. :)

eladb commented:
Allow both options sounds good to me.

A few pointers: the way a CDK app can communicate something like this to the toolkit is by attaching CDK metadata (`this.node.addMetadata`) on the stack object, and then the toolkit can look that up and apply the tags upon create/update of the stack.

I wonder if we should generalize the `Tag` aspect such that instead of [checking if a node is a `CfnResource`](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/cdk/lib/tag-aspect.ts#L74) it will just check if the implements `ITaggble` or something like that, and then each node can provide it's own tagging strategy (resources will add the tag to the CFN props and stacks will add metadata that will tell the toolkit to apply).

The owner question is: if I apply a tag at the stack level, and we know CFN will automatically apply them to all taggable resources, maybe the propagation should work differently.

@moofish32, thoughts?

IsmaelMartinez commented:
Waiting for @moofish32 answer but updated the PR with part of the @eladb requirements.

The code changes now does the following:

* Checks if arguments are passed for tags.
* If no arguments found, checks if there are Tags in the stack metadata.

I have tested that you can add stack tags with the following (been stack a cdk.Stack):
```
stack.node.addMetadata('Tags', [new cdk.Tag("Team", "Isma"), new cdk.Tag("Group", "fdsajkl3")]);
```

Note: should I change the PR title from parameters to arguments? Might be better wait until there is a more clear knowledge of all the things this does.

eladb commented on packages/aws-cdk/lib/api/cxapp/stacks.ts:
```diff
@@ -225,6 +226,22 @@ export class AppStacks {
+    for (const id of Object.keys(stack.metadata)) {
+      const metadata = stack.metadata[id];
+      for (const entry of metadata) {
+        switch (entry.type) {
```
What's wrong with an "if"?

IsmaelMartinez replied:
Me bad. I will change this today when I got sometime. (many meetings). Thanks for the review.

eladb commented on packages/aws-cdk/lib/api/cxapp/stacks.ts:
```diff
@@ -225,6 +226,22 @@ export class AppStacks {
+      const metadata = stack.metadata[id];
+      for (const entry of metadata) {
+        switch (entry.type) {
+          case "Tags":
```
Define the key in @aws-cdk/cx-api and use a "aws:cdk" prefix.

IsmaelMartinez replied:
Does the "aspect/no aspect" question have any implication on this? Happy to change this (after the many meetings I got today)

eladb commented on packages/aws-cdk/lib/api/cxapp/stacks.ts:
```diff
@@ -373,3 +390,13 @@ export interface SelectedStack extends cxapi.SynthesizedStack {
   originalName: string;
 }
+
+export type Tags = Tag[];
```
```ts
interface Tag {
  readonly key;
  readonly value;
}
```

Should suffice.

IsmaelMartinez commented:
> I am expecting the standard `Tag` aspect to work seamlessly:
> 
> ```ts
> const s = new MyStack(...);
> s.node.apply(new cdk.Tag('key, value'));
> ```

Agree but I was waiting for confirmation from @moofish32. This was just to confirm I could create a stack, add the "Tag" metadata and that will pass to the CDK part (so a WIP) 

As far as I understand, by adding the tags on the stack, there is no need to add the tags using the aspects. Does adding both adds a benefit that I don't see?

IsmaelMartinez commented:
I will see if I can dig around the aspects tomorrow, but only if we are really going to use them.

If, by tagging the stack, we tag all the resources, there might be a reason to keep them separate? Just wondering if that is a valid reasoning or we just tag everything (stack and resources with the aspects)

Thanks and apologies for my ignorance on this project, 1st typescript work and 1st AWS PR.

IsmaelMartinez commented:
Hi @eladb ,

Could you check my previous message? I am not sure why we will need to be able to do both types of tagging (all stack vs all the resources minus the stack).

Maybe I am missing something in here. I will see if I can have a look at the tag aspects today.

Soon I will be on holidays for a few weeks so I will like to wrap this beforehand. 

Any news also from @moofish32 ?

moofish32 commented:
Sorry for my delay. The reason I prefer the aspect over the CLI is I don't really want to type out all the tags for the stack on the command line or in bash via Jenkins. Aspects let me commit that to VCS. If the next thing we want is to override that for specific environments the CDK has many methods for that (e.g. context, config object, or json/yaml prop file). 

Now to @eladb's point a stack does not currently implement `ITaggable`, but that's very little code. I would think we just let aspects behave normally and in the [collectMetadata](https://github.com/awslabs/aws-cdk/blob/master/packages/@aws-cdk/cdk/lib/stack.ts#L545) we just render the tags from the TagManager. The tag format seems to be little different `Key=Value` we could just add a formatter for stack type. I did test if the resource has a tag and the stack has the same tag no errors or mistakes occurred for my simple security group test.

We might want to move the tag interface out of CfnResource and on to CfnElement?

IsmaelMartinez commented:
Hi @eladb and @moofish32 

Will need some tidy up, but code is not that far from the final solution (IMO). I will appreciate a bit of feedback to see if i am going in the right direction.

I have added a few changes but still an early WIP. Prefer to see it functionally working and then do the tests rather than writing all up and then needing to redo it.

The current version does support passing the tags as tag properties, applying the tag aspect and passing it with yargs.

To notice, if passing via yargs (--tags key=value) that then ignores the aspects one.

So now is possible to:

```
const app = new App();

const stack1 = new Stack(app, 'stack1', { tags: { foo: 'bar' } });
const stack2 = new Stacl(app, 'stack2');

stack1.node.apply(new Tag('fii', 'bug'));
stack2.node.apply(new Tag('boo', 'bug'));
```
That will produce 
* stack1 with tags `foo bar` and `fii bug`
* stack2 with tags `boo bug`

and 
```
cdk deploy --tags foo=bar
```
I have also updated a bit the code (pull from master a couple of days) so hopefully the integration is easier. 

Thanks

moofish32 commented on packages/aws-cdk/lib/api/cxapp/stacks.ts:
```diff
@@ -375,3 +396,8 @@ export interface SelectedStack extends cxapi.SynthesizedStack {
   originalName: string;
 }
+
+export interface Tag {
```
@rix0rrr -- we are still separating interfaces across the cxapi right?

IsmaelMartinez replied:
@rix0rrr , any updates on this?

eladb replied:
Why do we need this as a public API?

moofish32 commented on packages/aws-cdk/lib/settings.ts:
```diff
@@ -222,6 +212,42 @@ export class Settings {
     return ret;
   }
 
+  private static parseStringContextListToObject(argv: yargs.Arguments) {
```
@eladb - what's the position on private static?

IsmaelMartinez replied:
happy to remove the static if required...

moofish32 replied:
Mostly in the code base these types of helper like functions are just defined as `function( ... ) { }` like you see `expandHomeDir()` at the bottom of this class. I honestly don't know if that is a specific decision by the team or just what I've seen.

IsmaelMartinez replied:
Ok, the reason is because the fromCommandLineArguments is Static, so those functions need to be static. I will specify the return type on those in the next commit.

IsmaelMartinez commented on packages/@aws-cdk/cdk/lib/stack.ts:
```diff
@@ -153,6 +157,13 @@ export class Stack extends Construct {
     this.logicalIds = new LogicalIDs(props && props.namingScheme ? props.namingScheme : new HashedAddressingScheme());
     this.name = props.stackName !== undefined ? props.stackName : this.calculateStackName();
     this.autoDeploy = props && props.autoDeploy === false ? false : true;
+    if (props.tags) {
```
To be able to do this it needs to be of the type NotTaggable... I hope that is ok.

IsmaelMartinez replied:
I have made that change. Also tidied up a bit the code by adding a new formatter in the aspect

IsmaelMartinez commented:
> A couple of comments, and questions that I think we need @eladb or @rix0rrr to help with.

Hi @moofish32. I have made the only change that you requested. 

Once @eladb and/or @rix0rrr agree with the structure/logic , I will be happy to write some unit tests around the changes. Might need to wait until the end of the month as I am not going to be much available from next week. 

Thanks!

moofish32 commented on packages/@aws-cdk/cdk/lib/tag-aspect.ts:
```diff
@@ -104,7 +109,7 @@ export class Tag extends TagBase {
 
   protected applyTag(resource: ITaggable) {
-    if (resource.tags.applyTagAspectHere(this.props.includeResourceTypes, this.props.excludeResourceTypes)) {
+    if (resource.tags.applyTagAspectHere(this.props.includeResourceTypes, this.props.excludeResourceTypes) || Stack.isStack(resource)) {
```
Actually looking at this again, do you need `|| Stack.isStack(resource)` ? I think if you remove that everything should still work or am I missing something?

IsmaelMartinez replied:
@moofish32 , Just commit a change with this. 

I had to move the isTaggable up to the Contructor as, otherwise, I had to duplicate it in the "Stack" area.

IsmaelMartinez commented:
Hi @eladb , 

Because this changes affect the client, the integration tests will need to change. 

How/Who do I need to contact for this? 

Overall, I don't have anything else to change. Maybe a couple of small unit tests but for areas that aren't currently tested.

I might have some time tomorrow and friday to have a look into this.

IsmaelMartinez commented:
Hi @eladb , I am back. Any updates?

IsmaelMartinez commented on packages/@aws-cdk/cdk/lib/cfn-resource.ts:
```diff
@@ -264,6 +250,7 @@ export enum TagType {
   Standard = 'StandardTag',
   AutoScalingGroup = 'AutoScalingGroupTag',
   Map = 'StringToStringMap',
+  KeyValue = 'KeyValue',
```
They are slightly different. The StringToStringMap formats objects with `{[key:string] = value:string}`, while the KeyValueFormatter formats objects with StackTag type, that are `{Key: key, Value: value}` . 

If we want to use the same StringToStringMap then we need another formatting somewhere else, but thought better doing it in this formatTags. Let me know if that formatter is better move somewhere else.

eladb replied:
Okay, this is fine.

IsmaelMartinez commented on packages/@aws-cdk/cdk/lib/construct.ts:
```diff
@@ -573,6 +573,13 @@ export class Construct implements IConstruct {
+  /**
+   * Check whether the given construct is Taggable
+   */
+  public static isTaggable(construct: any): construct is ITaggable {
```
I have put it in the constructor as otherwise I need to add it in a few more places. 

The Stack is not a cfnRersouce (nor I think it should be).

If I independently add it to the Stack itself, then it gets pretty messy on the validations of objects. 

If you prefer, I am happy to create a "TaggableConstructor" that extend the Constructor and has the tags in there?

eladb replied:
Do we even need this as a public API? Sounds like it could be a helper function that we just put near the code that consumes it

IsmaelMartinez replied:
This has just been shift/lift from the cfn-resource.ts. Changing this might imply a fairly big refactoring on the TagAspect. Probably @moofish32 can give some guidance on this, or implement it.

eladb replied:
Still, `isTaggable` should not be on `Construct`. How about `TagManager.isTaggable`?

IsmaelMartinez replied:
Addressed in f0a219e

IsmaelMartinez commented on packages/@aws-cdk/cdk/lib/tag-manager.ts:
```diff
@@ -142,6 +146,33 @@ class MapFormatter implements ITagFormatter {
   }
 }
 
+class KeyValueFormatter implements ITagFormatter {
```
As in my previous comment. The format is slightly different.

IsmaelMartinez commented on packages/@aws-cdk/cdk/lib/tag-aspect.ts:
```diff
@@ -71,16 +74,19 @@ export abstract class TagBase implements IAspect {
-    if (CfnResource.isTaggable(resource)) {
-      this.applyTag(resource);
+    const resource = isCfnResource ? construct as CfnResource : construct as Stack;
+    if (Construct.isTaggable(resource)) {
```
redundant code from previous commits. 

I will resolve the conversation once I finish with the other changes/build. 

It should only need the last 3 lines, basically... if isTaggable apply the tags.

IsmaelMartinez commented on packages/@aws-cdk/cdk/lib/stack.ts:
```diff
@@ -153,6 +157,8 @@ export class Stack extends Construct {
     this.logicalIds = new LogicalIDs(props && props.namingScheme ? props.namingScheme : new HashedAddressingScheme());
     this.name = props.stackName !== undefined ? props.stackName : this.calculateStackName();
     this.autoDeploy = props && props.autoDeploy === false ? false : true;
+    const tags = props === undefined ? undefined : props.tags;
```
it should be `props.tags`. resolving once I get the merge from master, changes commit in and build green .

IsmaelMartinez commented on packages/@aws-cdk/cdk/lib/stack.ts:
```diff
@@ -72,6 +74,8 @@ export class Stack extends Construct {
 
   private static readonly VALID_STACK_NAME_REGEX = /^[A-Za-z][A-Za-z0-9-]*$/;
 
+  public readonly tags: TagManager;
```
By implementing the ITaggable, seems like I need to implement those tags (actually the TagManager) as public.

eladb replied:
OK

IsmaelMartinez commented:
All reviewed and ready for next phase of checks @eladb

eladb commented on packages/@aws-cdk/cdk/lib/cfn-resource.ts:
```diff
@@ -212,7 +198,7 @@ export class CfnResource extends CfnRefElement {
     try {
       // merge property overrides onto properties and then render (and validate).
-      const tags = CfnResource.isTaggable(this) ? this.tags.renderTags() : undefined;
+      const tags = Construct.isTaggable(this) ? this.tags.renderTags() : undefined;
```
As requested above, please remove `isTaggable` from `Construct.

IsmaelMartinez replied:
f0a219e

eladb commented on packages/@aws-cdk/cdk/lib/stack.ts:
```diff
@@ -40,6 +39,8 @@ export interface StackProps {
    */
   readonly autoDeploy?: boolean;
+
+  readonly tags?: { [key: string]: string };
```
Change "@default" to:

```
@default - No tags
```

eladb commented on packages/@aws-cdk/cx-api/lib/cxapi.ts:
```diff
@@ -31,6 +31,13 @@ export const DISABLE_VERSION_REPORTING = 'aws:cdk:disable-version-reporting';
 /**
+ * Tag metadata key.
+ */
+export const TAGS_METADATA_KEY = 'aws:cdk:stack-tags';
```
This needs to be in `metadata.ts` and rename to `STACK_TAGS_METADATA_KEY`

IsmaelMartinez replied:
0e3c101

eladb commented on packages/aws-cdk/lib/api/cxapp/stacks.ts:
```diff
@@ -62,6 +62,7 @@ export interface AppStacksProps {
  * In a class because it shares some global state
  */
 export class AppStacks {
+
```
Revert

IsmaelMartinez replied:
0e3c101

eladb commented on packages/aws-cdk/lib/api/deploy-stack.ts:
```diff
@@ -73,7 +75,8 @@ export async function deployStack(options: DeployStackOptions): Promise<DeploySt
     RoleARN: options.roleArn,
-    Capabilities: [ 'CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM', 'CAPABILITY_AUTO_EXPAND' ]
+    Capabilities: [ 'CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM', 'CAPABILITY_AUTO_EXPAND' ],
+    Tags: options.tags
```
What happens if this gets `undefined`?

IsmaelMartinez replied:
Probably one for integration. I have manually tested and means no tags. Meaning, any previous tags applied will be removed (what I think makes sense)

If no tags are provided by either way, via aspects or arguments, the cloudformation template doesn't populate the "Tags" (at least from the synth and the one in the aws cloudformation template output).

IsmaelMartinez commented:
With regards https://github.com/awslabs/aws-cdk/pull/2185#discussion_r289706234 , the Tag interface is used in the stacks.ts, deploy-stack.ts and deployment-target.ts. 

I consider that stacks.ts is the most suitable place for it, but let me know if there is another component that should hold this. Thanks!!

IsmaelMartinez commented:
ok, I have re-tested all of this and works as I will expect. (there was a wee issue after the code review changes that is now fixed)

Now it will be possible to:

```
const app = new App();

const stack1 = new Stack(app, 'stack1', { tags: { foo: 'bar' } });
const stack2 = new Stacl(app, 'stack2');

stack1.node.apply(new Tag('fii', 'bug'));
stack2.node.apply(new Tag('boo', 'bug'));
```
That will produce 
* stack1 with tags `foo bar` and `fii bug`
* stack2 with tags `boo bug`

It is possible also to override constructor tags with the stack.node.apply. 

So doing:
```
stack1.node.apply(new Tag('foo', 'newBar');
```
stack1 will have tags `foo newBar` and `fii bug`

Last, but not least, it is also possible to pass it via arguments (using yargs) as in the following example:

```
cdk deploy --tags foo=bar --tags myTag=myValue
```
That will produce a stack with tags `foo bar`and `myTag myValue`

**Important**
That will ignore tags provided by the constructor and/or aspects.

AlexRex commented:
After adding this into the library we cannot deploy Tags anymore. Steps to reproduce: 

With version:

```bash
$ cdk --version
0.34.0 (build 523807c)
```

Create new sample App:
```
$ cdk init sample-app --language=typescript
```

Add tags: 

```
#!/usr/bin/env node
import cdk = require('@aws-cdk/cdk');
import { CdkHacksStack } from '../lib/cdk-hacks-stack';

const app = new cdk.App();
new CdkHacksStack(app, 'CdkHacksStack', {
  env: {
    region: 'eu-west-1'
  },
  tags: {
    hack: 'nope!'
  }
});
```

Run deploy:
```
$ npx cdk deploy
```

Result:

```
 ❌  CdkHacksStack failed: MultipleValidationErrors: There were 3 validation errors:
* MissingRequiredParameter: Missing required key 'Key' in params.Tags[0]
* MissingRequiredParameter: Missing required key 'Value' in params.Tags[0]
* UnexpectedParameter: Unexpected key '0' found in params.Tags[0]
There were 3 validation errors:
* MissingRequiredParameter: Missing required key 'Key' in params.Tags[0]
* MissingRequiredParameter: Missing required key 'Value' in params.Tags[0]
* UnexpectedParameter: Unexpected key '0' found in params.Tags[0]
```

I've been debugging a bit and it looks like in the class `param_validator` the Tags object is an array of arrays `[ [ { Key: 'hack', Value: 'nope!' } ] ]`. I guess this should expect a simply array. 

The unique way I found this PR working is when passing the parameters in the CLI.

eladb commented:
https://github.com/IsmaelMartinez can you take a look?

IsmaelMartinez commented:
Oh dear, 1st of all apologies. I will see if I can have a look at it tonight if not tomorrow morning

skinny85 commented:
@AlexRex I see you're running `cdk --version`, but then `npx cdk deploy`. Can you make sure both CDK CLI versions (the one you used for `init`, and the one you used for `deploy`) are in the same version?

IsmaelMartinez commented:
I will continue investigation and others in the open issue.

AlexRex commented:
@skinny85 is the same version, tested out. 

@IsmaelMartinez no worries, I just following the other issue. Thanks!

iDVB commented:
Is there docs on what the net-net way is to implement these tags?
This thread jumped around a bit and I didn't see the Guide being updated as part of this merge?
